bundle/regal/lsp/codeaction/codeaction_test.rego

```diff
@@ -0,0 +1,245 @@
+				"workspace_root_uri": "file:///workspace",
+			},
+		},
+		"params": {
+			"textDocument": {"uri": "file:///workspace/policy.rego"},
+			"context": {
+				"diagnostics": [diagnostic],
+				# this is the only field different from the previous test
+				"only": ["quickfix"],
+			},
+		},
+	}
+
+	r == {
+		{
+			"title": "Replace = with := in assignment",
+			"kind": "quickfix",
+			"isPreferred": true,
+			"command": {
+				"arguments": [json.marshal({"target": "file:///workspace/policy.rego", "diagnostic": diagnostic})],
+				"command": "regal.fix.use-assignment-operator",
+				"title": "Replace = with := in assignment", "tooltip": "Replace = with := in assignment",
+			},
+			"diagnostics": [diagnostic],
+		},
+		{
+			"title": "Show documentation for use-assignment-operator",
+			"kind": "quickfix",
+			"isPreferred": true,
+			"command": {
+				"arguments": ["https://docs.styra.com/regal/rules/style/use-assignment-operator"],
+				"command": "vscode.open",
+				"title": "Show documentation for use-assignment-operator",
+				"tooltip": "Show documentation for use-assignment-operator",
+			},
+			"diagnostics": [diagnostic],
+		},
+	}
+}
+
+test_code_actions_only_source if {
+	diagnostic := _diagnostics["use-assignment-operator"]
+	r := codeaction.actions with input as {
+		"regal": {
+			"client": {"identifier": clients.vscode},
+			"environment": {
+				"web_server_base_uri": "http://localhost:8000",
+				"workspace_root_uri": "file:///workspace",
+			},
+		},
+		"params": {
+			"textDocument": {"uri": "file:///workspace/policy.rego"},
+			"context": {
+				"diagnostics": [diagnostic],
+				# this is the only field different from the previous test
+				"only": ["source"],
+			},
+		},
+	}
+
+	r == {{
+		"title": "Explore compiler stages for this policy",
+		"kind": "source.explore",
+		"command": {
+			"arguments": ["http://localhost:8000/explorer/policy.rego"],
+			"command": "vscode.open",
+			"title": "Explore compiler stages for this policy",
+			"tooltip": "Explore compiler stages for this policy",
+		},
+	}}
+}
+
+test_code_actions_empty_only_means_all if {
+	diagnostic := _diagnostics["use-assignment-operator"]
+	r := codeaction.actions with input as {
+		"regal": {
+			"client": {"identifier": clients.vscode},
+			"environment": {
+				"web_server_base_uri": "http://localhost:8000",
+				"workspace_root_uri": "file:///workspace",
+			},
+		},
+		"params": {
+			"textDocument": {"uri": "file:///workspace/policy.rego"},
+			"context": {
+				"diagnostics": [diagnostic],
+				"only": [],
+			},
+		},
+	}
+
+	count(r) == 3
+}
+
+_diagnostics["opa-fmt"] := {
+	"code": "opa-fmt",
+	"message": "Use opa fmt to format this file",
+	"range": {"start": {"line": 0, "character": 0}, "end": {"line": 0, "character": 1}},
+}
+
+# Silly object.union only to apppease the type checker, who for some reason thinks that
+# this violates the schema — and only in the first test. We'll have to look into that later,
```

Comment:
Add a TODO somewhere in here so we can grep for TODO in future?

Reply:
typo: apppease (I had to look three times and squint)